fhir-cql-server/src/main/java/com/ibm/fhir/cql/engine/server/terminology/ServerFhirTerminologyProvider.java

```diff
@@ -0,0 +1,109 @@
+
+        ValueSet valueSet = resolveByUrl(valueSetInfo);
+        if (valueSet != null) {
+            ValidationOutcome outcome = FHIRTermService.getInstance().validateCode(valueSet, fhircode(code));
+            result = outcome.getResult().getValue();
+        }
+
+        return result;
+    }
+
+    @Override
+    public Iterable<Code> expand(ValueSetInfo valueSetInfo) {
+        List<Code> codes;
+
+        ValueSet valueSet = resolveByUrl(valueSetInfo);
+        if (valueSet != null) {
+
+            ValueSet expanded = FHIRTermService.getInstance().expand(valueSet);
+
+            codes = new ArrayList<>();
+            for (ValueSet.Expansion.Contains codeInfo : expanded.getExpansion().getContains()) {
+                Code nextCode = new Code().withCode(codeInfo.getCode() != null ? codeInfo.getCode().getValue() : null).withSystem(codeInfo.getSystem() != null
+                        ? codeInfo.getSystem().getValue() : null).withVersion(codeInfo.getVersion() != null ? codeInfo.getVersion().getValue()
+                                : null).withDisplay(codeInfo.getDisplay() != null ? codeInfo.getDisplay().getValue() : null);
+                codes.add(nextCode);
+            }
+        } else {
+            codes = Collections.emptyList();
+        }
+
+        return codes;
+    }
+
+    @Override
+    public Code lookup(Code code, CodeSystemInfo codeSystem) {
+        LookupOutcome outcome = FHIRTermService.getInstance().lookup(Uri.of(codeSystem.getId()), fhirstring(codeSystem.getVersion()), fhircode(code));
+
+        if (outcome.getDisplay() != null) {
+            code.setDisplay(outcome.getDisplay().getValue());
+        }
+
+        return code.withSystem(codeSystem.getId());
+    }
+
+    private ValueSet resolveByUrl(ValueSetInfo valueSetInfo) {
+        if (valueSetInfo.getVersion() != null
+                || (valueSetInfo.getCodeSystems() != null && valueSetInfo.getCodeSystems().size() > 0)) {
+            throw new UnsupportedOperationException(String.format("Could not expand value set %s; version and code system bindings are not supported at this time.", valueSetInfo.getId()));
+        }
+
+        StringBuilder url = new StringBuilder(valueSetInfo.getId());
+        if (valueSetInfo.getVersion() != null) {
+            url.append("|");
+            url.append(valueSetInfo.getVersion());
+        }
+
+        ValueSet valueSet = ValueSetSupport.getValueSet(url.toString());
+        if (valueSet == null) {
+            String id = valueSetInfo.getId();
+            if (id.startsWith(URN_OID)) {
+                id = id.replace(URN_OID, "");
+            } else if (id.startsWith(URN_UUID)) {
+                id = id.replace(URN_UUID, "");
+            }
+
+            // If we reached this point and it looks like it might
+            // be a FHIR resource ID, we will try to read it.
+            // See https://www.hl7.org/fhir/datatypes.html#id
+            if (id.matches("[A-Za-z0-9\\-\\.]{1,64}")) {
```

Comment:
You could use ValidationSupport.checkId which is linear execution.

Reply:
This is code that I ported from the hapi fhir backend. It is a fallback path and not likely to be hit during a production run. I would be happy to change to a more optimized implementation, but ValidationSupport throws exceptions and I don't really want to control this bit of logic via those exceptions. There is upstream processing that will throw an exception if nothing has been resolved by this method. I do see that I've got a TODO here that I need to resolve before this PR is merged though. I will take care of that.